File: Final_SVM_s.py
# ======================================================================
#  ECG_Auth_System.py      rev-I  (2025-04-24)
#  FINAL_SVM_S  (feature-based multiclass SVM)
# ======================================================================

import os, re
import numpy as np
import pywt
from scipy.signal import butter, filtfilt, find_peaks, welch, resample_poly
from sklearn.svm import SVC
from sklearn.multiclass import OneVsRestClassifier

# -------------------- globals (mirror MATLAB "global") --------------------
SVMModel = None
mima = None
Xtrain_mean = None
Xtrain_std = None


# =================  辅助函数   =================
def resample_to(x, fs0, fs1):
    if fs0 == fs1:
        return np.asarray(x, dtype=np.float64)
    from fractions import Fraction
    frac = Fraction(fs1, fs0).limit_denominator(2000)
    return resample_poly(np.asarray(x, dtype=np.float64), frac.numerator, frac.denominator).astype(np.float64)


# 降噪 Denoising
def denoiseECG(x, Fs):
    low, high = 0.5, 40.0
    b, a = butter(4, [low/(Fs/2), high/(Fs/2)], btype="band")
    y = filtfilt(b, a, x).astype(np.float64)
    # y = (y - mean(y)) ./ std(y);   % MATLAB commented out
    return y


# —— 提取基本特征 extract basic features—— %
def extractFeatures(ecg, Fs):
    ecg = np.asarray(ecg, dtype=np.float64)

    # —— 时域特征 time-domain features—— %
    loc, _ = find_peaks(
        ecg,
        height=(np.mean(ecg) + 0.5*np.std(ecg)),
        distance=max(1, int(round(0.25*Fs)))
    )

    RR = np.diff(loc) / Fs if loc.size >= 2 else np.array([])
    if RR.size < 3:
        RR = np.array([np.nan, np.nan, np.nan], dtype=np.float64)

    feat_meanRR = np.nanmean(RR)
    feat_SDNN   = np.nanstd(RR)
    feat_RMSSD  = np.sqrt(np.nanmean(np.diff(RR)**2)) if RR.size >= 2 else np.nan

    # —— 波形相似度 similarity between waveforms—— %
    beatLen = int(round(0.7 * Fs))
    if loc.size >= 2:
        half = beatLen // 2
        tmpl = ecg[max(0, loc[0]-half):min(ecg.size, loc[0]+half)]
        nCmp = min(8, loc.size - 1)
        ccs = []
        for k in range(nCmp):
            seg = ecg[max(0, loc[k+1]-half):min(ecg.size, loc[k+1]+half)]
            L = min(len(tmpl), len(seg))
            if L >= 10 and np.std(tmpl[:L]) > 0 and np.std(seg[:L]) > 0:
                ccs.append(np.corrcoef(tmpl[:L], seg[:L])[0, 1])
        feat_shape = float(np.nanmean(ccs)) if len(ccs) else np.nan
    else:
        feat_shape = np.nan

    # —— 频域HRV特征 frequency-domain HRV features—— %
    f, pxx = welch(ecg, fs=Fs, window="hamming", nperseg=512, noverlap=256, nfft=1024, scaling="density")

    def bandpower(bnd):
        lo, hi = bnd
        m = (f >= lo) & (f <= hi)
        if not np.any(m):
            return 0.0
        return float(np.trapz(pxx[m], f[m]))

    feat_LF = bandpower((0.04, 0.15))
    feat_HF = bandpower((0.15, 0.40))
    feat_LFHF = feat_LF / max(feat_HF, np.finfo(np.float64).eps)

    vec = np.round(
        np.array([feat_meanRR, feat_SDNN, feat_RMSSD, feat_shape, feat_LF, feat_HF, feat_LFHF], dtype=np.float64),
        3
    )

    return {
        "meanRR": feat_meanRR,
        "SDNN": feat_SDNN,
        "RMSSD": feat_RMSSD,
        "shape": feat_shape,
        "LF": feat_LF,
        "HF": feat_HF,
        "LFHF": feat_LFHF,
        "vec": vec
    }


# —— generate a hashed password —— %
def generatePassword(vec):
    s = "_".join([f"{float(x):.3f}" for x in np.asarray(vec).ravel().tolist()])
    import hashlib
    return hashlib.sha256(s.encode("utf-8")).hexdigest().lower()


# ---------- 文件读取 file reading ----------
def loadAmpECG_manual(folder, rec, lead):
    hea_path = os.path.join(folder, rec + ".hea")
    dat_path = os.path.join(folder, rec + ".dat")

    with open(hea_path, "r", errors="ignore") as f:
        lines = [ln.strip() for ln in f.read().splitlines() if ln.strip()]
    lines = [ln for ln in lines if not ln.startswith("#")]

    hdr = re.split(r"\s+", lines[0])
    nSig = int(float(hdr[1]))
    Fs = float(hdr[2])  # Sampling rate 采样率

    rows = lines[1:]
    rows = [ln for ln in rows if "filtered" not in ln.lower()]

    lead = max(1, min(int(lead), len(rows)))  # 1-based
    tok = re.split(r"\s+", rows[lead-1])

    # MATLAB uses tok{3} and tok{6} (1-based) => tok[2], tok[5]
    gain = float(tok[2])
    baseline = float(tok[5])

    raw = np.fromfile(dat_path, dtype="<i2").astype(np.float64)
    L = (raw.size // nSig) * nSig
    raw = raw[:L]
    sig = raw.reshape(nSig, -1)[lead-1, :]

    if gain != 0:
        sig = (sig - baseline) / gain

    # 小波处理 wavelet processing
    wname = "coif4"
    level = 7
    coeffs = pywt.wavedec(sig, wname, level=level)
    coeffs[0] = np.zeros_like(coeffs[0])  # A7=0
    sig = pywt.waverec(coeffs, wname)

    return sig.astype(np.float64), Fs


# =================== AMP 分组初始化 ===================
def initUsersFromAMP(ampDir, Fs_target, band, leadIdx, winSec, db):
    heaFiles = [f for f in os.listdir(ampDir) if f.endswith(".hea")]
    recNames = sorted(list({os.path.splitext(f)[0] for f in heaFiles}))
    groups = {}

    for rec in recNames:
        m = re.match(r"^(.*?_\d+)_\d+$", rec)
        if not m:
            continue
        key = m.group(1)
        groups.setdefault(key, []).append(rec)

    keys = list(groups.keys())

    Xtrain = []
    labels = []
    userID = 1

    global mima, Xtrain_mean, Xtrain_std, SVMModel
    mima = []

    for n in range(0, max(0, len(keys) - 1)):
        recSet = groups[keys[n]]
        user_feats = []  # 每个用户单独积

        for r in recSet[0:len(recSet)]:
            ecg, Fs0 = loadAmpECG_manual(ampDir, r, leadIdx)
            if Fs0 != Fs_target:
                ecg = resample_to(ecg, Fs0, Fs_target)
            ecg = denoiseECG(ecg, Fs_target)

            step = int(Fs_target * winSec)

            if len(ecg) > 100000:
                for start_idx in range(0, len(ecg) - step, step):
                    ecg_seg = ecg[start_idx:start_idx + step]
                    feat = extractFeatures(ecg_seg, Fs_target)
                    vec = feat["vec"]

                    # 检查并处理NaN Check and handle NaN values
                    if np.any(np.isnan(vec)):
                        continue

                    Xtrain.append(vec)
                    labels.append(userID)
                    user_feats.append(vec)
            else:
                ecg_seg = ecg[:min(len(ecg), step)]
                if ecg_seg.size < step:
                    ecg_seg = np.pad(ecg_seg, (0, step - ecg_seg.size))
                feat = extractFeatures(ecg_seg, Fs_target)
                vec = feat["vec"]

                if np.any(np.isnan(vec)):
                    continue

                Xtrain.append(vec)
                labels.append(userID)
                user_feats.append(vec)

        # 生成密码 Generate password
        if len(user_feats) > 0:
            mean_feat = np.mean(np.vstack(user_feats), axis=0)
            pwd = generatePassword(mean_feat)
            mima.append(pwd)
            print(f"[USER] Registered user {userID}  ({len(user_feats)} segments)")
            userID += 1
        else:
            print(f"[WARN] No valid segments for user {userID}, skipping.")

    Xtrain = np.vstack(Xtrain).astype(np.float64)
    labels = np.array(labels, dtype=int)

    print(f"[CHECK] Final Xtrain size: {Xtrain.shape[0]} x {Xtrain.shape[1]}")
    print(f"[CHECK] Number of NaN entries: {np.isnan(Xtrain).sum()}")

    if Xtrain.shape[0] == 0:
        raise RuntimeError("Xtrain is empty! Initialization failed.")

    # 标准化训练数据 Standardize training data before training the model
    Xtrain_mean = Xtrain.mean(axis=0)
    Xtrain_std = Xtrain.std(axis=0)
    Xtrain_std[Xtrain_std == 0] = 1.0
    Xtrain = (Xtrain - Xtrain_mean) / Xtrain_std

    # MATLAB: fitcecoc(Xtrain, labels)
    SVMModel = OneVsRestClassifier(SVC(kernel="rbf", gamma="scale"))
    SVMModel.fit(Xtrain, labels)

    db["users"] = [{"id": i+1} for i in range(len(mima))]
    return db


def Train_data(ampDir, Fs_target, band, leadIdx, winSec, db):
    # 测试阶段 (使用提取的特征识别用户)  Testing phase (identify users using extracted features)

    heaFiles = [f for f in os.listdir(ampDir) if f.endswith(".hea")]
    recNames = sorted(list({os.path.splitext(f)[0] for f in heaFiles}))
    groups = {}

    for rec in recNames:
        m = re.match(r"^(.*?_\d+)_\d+$", rec)
        if not m:
            continue
        key = m.group(1)
        groups.setdefault(key, []).append(rec)

    keys = list(groups.keys())
    Xtest = []
    trueLabels = []
    userID = 1

    global Xtrain_mean, Xtrain_std, SVMModel, mima

    for n in range(0, len(keys)):
        recSet = groups[keys[n]]
        if len(recSet) == 0:
            continue

        # 只取每个人最后一条记录 only take the last record for each person。
        rec = recSet[-1]
        ecg, Fs0 = loadAmpECG_manual(ampDir, rec, leadIdx)
        if Fs0 != Fs_target:
            ecg = resample_to(ecg, Fs0, Fs_target)
        ecg = denoiseECG(ecg, Fs_target)

        # 提取最后20秒的特征 extract features from the last 20 seconds
        step = int(Fs_target * winSec)
        if len(ecg) > step:
            ecg_seg = ecg[-step:]
        else:
            ecg_seg = ecg[:min(len(ecg), step)]
            if ecg_seg.size < step:
                ecg_seg = np.pad(ecg_seg, (0, step - ecg_seg.size))

        feat = extractFeatures(ecg_seg, Fs_target)
        vec = feat["vec"]
        if np.any(np.isnan(vec)):
            print(f"[WARN] Test sample for user {userID} has NaN, skipping.")
            continue

        Xtest.append(vec)
        trueLabels.append(userID)
        userID += 1

    if len(Xtest) == 0:
        print("[WARN] Xtest is empty; no predictions can be made.")
        return

    Xtest = np.vstack(Xtest).astype(np.float64)

    # 标准化测试特征 standardize test features
    Xtest = (Xtest - Xtrain_mean) / Xtrain_std

    predictedLabels = SVMModel.predict(Xtest)

    # 打印结果 print results。
    for idx, (t, p) in enumerate(zip(trueLabels, predictedLabels), start=1):
        p = int(p)
        pwd = mima[p-1] if (mima is not None and 1 <= p <= len(mima)) else "<out-of-range>"
        print(f"[{idx:2d}] Known user #{t} -> Predicted user #{p}   pwd={pwd}")


def main():
    # =================== 全局参数 global parameters。===================
    Fs_target = 500
    winSec  = 20
    band      = (0.5, 40)
    batchN    = 4
    margin    = -0.02
    leadIdx   = 1
    ampDir    = "./amp_"
    testDir   = "./test"
    dbFile    = "ecg_db.mat"

    print("[INFO] ECG Auth System started")

    # =================== 0) 载入 / 新建数据库 Load / Create Database===================
    db = {"users": []}
    print("[INFO] Created new empty database")

    # =================== 1) AMP 预初始化 AMP pre-initialization ===================
    if (len(db["users"]) == 0) and os.path.isdir(ampDir):
        print("[INIT] Building initial users from AMP ...")
        db = initUsersFromAMP(ampDir, Fs_target, band, leadIdx, winSec, db)
        Train_data(ampDir, Fs_target, band, leadIdx, winSec, db)
        print(f"[INIT] Users after AMP init: {len(db['users'])}")
    else:
        print(f"[WARN] AMP dir not found or db not empty: ampDir={ampDir}")


if __name__ == "__main__":
    main()

File: Final_SVM_s.m
% ======================================================================
%  ECG_Auth_System.m      rev-I  (2025-04-24)
% ======================================================================
clc; clear; close all;

%% =================== 全局参数 global parameters。===================
Fs_target = 500;            % 统一采样率 uniform sampling rate
winSec  = 20;             % 窗口长度 (秒)window length (seconds)。
band      = [0.5 40];       % 滤波带宽 filter bandwidth。
batchN    = 4;              % 累积 N 段再增量更新 accumulate N segments before incremental update。
margin    = -0.02;           %  模糊带阈值  fuzzy band threshold。
leadIdx   = 1;              % AMP 取第 1 导联 Use lead I from AMP 
ampDir    = './amp_';        % 可选 AMP 训练目录
testDir   = './test';       % 测试 CSV 目录 test CSV directory
dbFile    = 'ecg_db.mat';   % 数据库文件 database file。


fprintf('[INFO] ECG Auth System started  %s\n', datestr(now,31));

%% =================== 0) 载入 / 新建数据库 Load / Create Database===================
if isfile(dbFile)&&0
    s  = load(dbFile,"db"); db = s.db;
    fprintf('[INFO] Loaded database (%d users)\n', numel(db.users));
else
    db.users = struct([]);
    save(dbFile,"db","-mat");
    fprintf('[INFO] Created new empty database\n');
end

%% =================== 1) AMP 预初始化 AMP pre-initialization ===================
if isempty(db.users) && isfolder(ampDir)
    fprintf('[INIT] Building initial users from AMP ...\n');
    db = initUsersFromAMP(ampDir, Fs_target, band, leadIdx, winSec, db);
    Train_data(ampDir, Fs_target, band, leadIdx, winSec, db);
    save(dbFile,"db","-mat");
    fprintf('[INIT] Users after AMP init: %d\n', numel(db.users));
end




%% ---------- AMP 分组初始化 ----------
function db = initUsersFromAMP(ampDir, Fs_target, band, leadIdx, winSec, db)

heaFiles = dir(fullfile(ampDir,'*.hea'));
recNames = unique( erase({heaFiles.name}, '.hea') );
groups   = containers.Map;

for f = recNames
    m = regexp(f{1},'^(.*?_\d+)_\d+$','tokens'); 
    if isempty(m), continue; end
    key = m{1}{1};
    if isKey(groups,key)
        tmp = groups(key); tmp{end+1} = f{1}; groups(key) = tmp;
    else
        groups(key) = {f{1}};
    end
end

keys = groups.keys;
Xtrain = [];
labels  = [];
userID = 1;
global mima
mima = [];

for n = 1:numel(keys)-1
    recSet = groups(keys{n});
    user_feats = [];  % 每个用户单独积

    for r = recSet(1:(length(recSet)))

        [ecg,Fs0] = loadAmpECG_manual(ampDir,r{1},leadIdx);
        if Fs0~=Fs_target
            ecg = resample(ecg,Fs_target,Fs0);
        end
        ecg = denoiseECG(ecg,Fs_target);
        
        if(length(ecg)>100000)
            for start_idx = 1 : Fs_target*winSec  : (length(ecg) - Fs_target*winSec )
                ecg_seg = ecg(start_idx : start_idx + Fs_target*winSec  - 1);
                feat = extractFeatures(ecg_seg, Fs_target);
                vec = feat.vec;

                % 检查并处理NaN Check and handle NaN values
                if any(isnan(vec))
                    continue;  % 直接跳过这段坏样本
                end

                Xtrain = [Xtrain; feat.vec];
                labels = [labels; userID];
                user_feats = [user_feats; vec];
            end
        else
            ecg_seg = ecg(1:min(length(ecg), Fs_target*winSec ));
            feat = extractFeatures(ecg_seg, Fs_target);
            vec = feat.vec;

            % 检查并处理NaN Check and handle NaN values
            if any(isnan(vec))
                continue;  % 直接跳过这段坏样本
            end

            Xtrain = [Xtrain; feat.vec];
            labels = [labels; userID];
            user_feats = [user_feats; vec];
        end
    end

    % 生成密码 enerate password
    if ~isempty(user_feats)
        mean_feat =  mean(user_feats, 1);
        pwd = generatePassword(mean_feat);
        mima(userID,:) = pwd;
        fprintf('[USER] Registered user %d  (%d segments)\n', userID, size(user_feats,1));
        userID = userID + 1;
    else
        fprintf('[WARN] No valid segments for user %d, skipping.\n', userID);
    end
end

% 检查最终Xtrain Check final Xtrain
fprintf('[CHECK] Final Xtrain size: %d x %d\n', size(Xtrain,1), size(Xtrain,2));
fprintf('[CHECK] Number of NaN entries: %d\n', sum(isnan(Xtrain(:))));

if isempty(Xtrain)
    error('Xtrain is empty! Initialization failed.');
end

% 标准化训练数据 Standardize training data before training the model
global Xtrain_mean Xtrain_std
Xtrain_mean = mean(Xtrain, 1);
Xtrain_std = std(Xtrain, [], 1);
Xtrain_std(Xtrain_std == 0) = 1;  % 防止除以0，导致NaN
Xtrain = (Xtrain - Xtrain_mean) ./ Xtrain_std;

global SVMModel
SVMModel = fitcecoc(Xtrain, labels);
% SVMModel = fitcecoc(Xtrain, labels, 'OptimizeHyperparameters', 'auto', ...
%     'HyperparameterOptimizationOptions', struct('AcquisitionFunctionName', ...
%     'expected-improvement-plus'));
end


function Train_data(ampDir, Fs_target, band, leadIdx, winSec, db)
% 测试阶段 (使用提取的特征识别用户)  Testing phase (identify users using extracted features)

heaFiles = dir(fullfile(ampDir,'*.hea'));
recNames = unique(erase({heaFiles.name}, '.hea'));
groups = containers.Map;

for f = recNames
    m = regexp(f{1},'^(.*?_\d+)_\d+$','tokens');
    if isempty(m), continue; end
    key = m{1}{1};
    if isKey(groups,key)
        tmp = groups(key); tmp{end+1} = f{1}; groups(key) = tmp;
    else
        groups(key) = {f{1}};
    end
end

keys = groups.keys;
Xtest = []; % 测试特征 test features。
trueLabels = []; % 测试时的真实标签 ground truth labels during testing
userID = 1;

global Xtrain_mean Xtrain_std SVMModel mima

for n = 1:numel(keys)
    recSet = groups(keys{n});
    if isempty(recSet)
        continue;
    end

    % 只取每个人最后一条记录 only take the last record for each person。
    [ecg, Fs0] = loadAmpECG_manual(ampDir, recSet{end}, leadIdx);
    if Fs0 ~= Fs_target
        ecg = resample(ecg, Fs_target, Fs0);
    end
    ecg = denoiseECG(ecg, Fs_target);

    % 提取最后20秒的特征 extract features from the last 20 seconds
    if length(ecg) > Fs_target*winSec
        ecg_seg = ecg(end - Fs_target*winSec + 1 : end);
    else
        ecg_seg = ecg(1:min(end, Fs_target*winSec));
    end

    feat = extractFeatures(ecg_seg, Fs_target);
    vec = feat.vec;
    if any(isnan(vec))
        fprintf('[WARN] Test sample for user %d has NaN, skipping.\n', userID);
        continue;
    end

    Xtest = [Xtest; feat.vec];
    trueLabels = [trueLabels; userID];

    userID = userID + 1;
end

% 标准化测试特征 standardize test features
Xtest = (Xtest - Xtrain_mean) ./ Xtrain_std;

% 预测prediction
[predictedLabels, scores] = predict(SVMModel, Xtest);

% 打印结果 print results。
for idx = 1:length(predictedLabels)
    fprintf('[%2d] Known user #%d -> Predicted user #%d   pwd=%s\n', ...
        idx, trueLabels(idx), predictedLabels(idx), mima(predictedLabels(idx),:));
end

end



%% ---------- 文件读取 file reading----------
function [sig, Fs] = loadAmpECG_manual(folder, rec, lead)
txt = fileread(fullfile(folder,[rec '.hea']));
lines = regexp(txt,'\r?\n','split'); lines = lines(~cellfun('isempty',lines));
lines = lines(~startsWith(lines,'#'));
hdr   = regexp(lines{1},'\s+','split');
nSig  = str2double(hdr{2}); 
Fs = str2double(hdr{3});%Sampling rate 采样率
rows  = lines(2:end); 
rows = rows(~contains(rows,'filtered','IgnoreCase',true));
lead  = min(lead,numel(rows));
tok   = regexp(rows{lead},'\s+','split');
gain  = str2double(tok{3}); baseline = str2double(tok{6});
fid = fopen(fullfile(folder,[rec '.dat']),'r','ieee-le');
raw = fread(fid, inf, 'int16=>double'); fclose(fid);
sig = reshape(raw, nSig, []); 
sig = sig(lead,:);
if gain ~= 0, sig = (sig - baseline)/gain; end
wname = 'coif4';    % 小波类型wavelet type。
level = 7;          % 分解尺度decomposition scale
[C, L] = wavedec(sig, level, wname); % 多尺度分解 multi-scale decomposition

a7_start = 1;                  % A7在C中的起始位置 Starting position of A7 in C。
a7_end = L(1);                 % A7在C中的结束位置 Ending position of A7 in C。
C_zeroA7 = C;                  % 复制系数 copy coefficients
C_zeroA7(a7_start:a7_end) = 0; % 将A7部分置零Set the A7 component to zero 

%% 4. 重构信号（A7=0）
sig = waverec(C_zeroA7, L, wname); % 重构信号 reconstructed signal。
end

function [sig, Fs] = loadCsvECG(fname, targetFs)
data = readmatrix(fname,'NumHeaderLines',13,'OutputType','double');
if isvector(data), raw = double(data(:));
else, raw = double(data(:,1)); end
sig = resample(raw.', targetFs, targetFs); Fs = targetFs;
end


%% =================  辅助函数   =================
% 降噪 Denoising
function y = denoiseECG(x, Fs)
d = designfilt('bandpassiir','FilterOrder',4, ...
    'HalfPowerFrequency1',0.5,'HalfPowerFrequency2',40, ...
    'SampleRate',Fs);
y = filtfilt(d, x);
%y = (y - mean(y)) ./ std(y);
end

% —— 提取基本特征 extract basic features—— %
function feat = extractFeatures(ecg, Fs)

% —— 时域特征 time-domain features—— %
[~,loc] = findpeaks(ecg, ...
    'MinPeakHeight', mean(ecg) + 0.5*std(ecg), ...
    'MinPeakDistance', round(0.25*Fs));  % 检测心电信号中的峰值 Detect peaks in ECG signals.
RR = diff(loc) / Fs;%RR间期 RR interval
if numel(RR) < 3, RR = nan(1,3); end
% 平均RR间期 平均每次心跳的间隔，代表基础心率 mean RR interval the baseline heart rate.
feat.meanRR = mean(RR, 'omitnan');
% RR间期标准差 表示心率波动大小（整体变异性）The standard deviation of RR intervals the magnitude of heart rate fluctuations (overall variability).
feat.SDNN   = std( RR, 'omitnan');
% 连续RR差分的均方根 主要反映短期心率变化  root mean square of successive differences (RMSSD) of RR intervals short-term heart rate variability.
feat.RMSSD  = rms( diff(RR), 'omitnan');

% —— 波形相似度 tsimilarity between waveforms—— %
beatLen = round(0.7 * Fs);
tmpl    = ecg( max(1,loc(1)-beatLen/2) : min(end,loc(1)+beatLen/2-1) );
nCmp    = min(8, numel(loc)-1);
ccs     = nan(1, nCmp);
for k = 1:nCmp
    seg = ecg( max(1,loc(k+1)-beatLen/2) : min(end,loc(k+1)+beatLen/2-1) );
    L   = min(numel(tmpl), numel(seg));
    if L >= 10
        R = corrcoef(tmpl(1:L), seg(1:L));
        ccs(k) = R(1,2);
    end
end
feat.shape = mean(ccs, 'omitnan');

% —— 频域HRV特征 frequency-domain HRV features—— %
% 计算功率谱密度 compute the power spectral density (PSD)
[pxx, f] = pwelch(ecg, hamming(512), 256, 1024, Fs);
% LF（0.04–0.15 Hz）功率
feat.LF   = bandpower(pxx, f, [0.04 0.15], 'psd');
% HF（0.15–0.40 Hz）功率
feat.HF   = bandpower(pxx, f, [0.15 0.40], 'psd');
% LF/HF 比值
feat.LFHF = feat.LF / max(feat.HF, eps);

% —— 返回行向量 return a row vector—— %
names = fieldnames(feat);
v     = cellfun(@(f) round(feat.(f), 3), names);
feat.vec = v(:)';
end

% —— generate a hashed password —— %
function pwd = generatePassword(vec)
s  = sprintf('%.3f_',vec);
s(end)=[];
md = java.security.MessageDigest.getInstance('SHA-256');
h  = md.digest(uint8(s));
pwd = lower(reshape(dec2hex(typecast(h,'uint8'))',1,[]));
end
